component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPTokenGenerator.java

```diff
@@ -69,285 +72,360 @@
+
+            sendEmailWithToken(username, authenticatedUser, tenantAwareUsername, username, secretKey,
+                    email, tenantDomain, context);
+        }
+    }
+
+    private static long sendEmailWithToken(String username, AuthenticatedUser authenticatedUser,
+                                           String tenantAwareUsername, String firstName, String secretKey,
+                                           String email, String tenantDomain, AuthenticationContext context)
+            throws TOTPException {
+
+        long token;
+        byte[] secretKeyByteArray;
+        try {
+            String encoding = TOTPUtil.getEncodingMethod(tenantDomain, context);
+            if (TOTPAuthenticatorConstants.BASE32.equals(encoding)) {
+                Base32 codec32 = new Base32();
+                secretKeyByteArray = codec32.decode(secretKey);
+            } else {
+                Base64 codec64 = new Base64();
+                secretKeyByteArray = codec64.decode(secretKey);
+            }
+            token = generateToken(username, tenantDomain, secretKeyByteArray, context);
+            // Check whether the authenticator is configured to use the event handler implementation.
+            if (TOTPUtil.isEventHandlerBasedEmailSenderEnabled()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("TOTP authenticator configured to use the event handler implementation.");
+                }
+                triggerEvent(authenticatedUser.getUserName(), authenticatedUser.getTenantDomain(), email,
+                        authenticatedUser.getUserStoreDomain(), TOTPAuthenticatorConstants.EVENT_NAME,
+                        Long.toString(token));
+            } else {
+                sendNotification(tenantAwareUsername, firstName, tenantDomain, Long.toString(token), email);
+            }
+            if (log.isDebugEnabled()) {
+                log.debug(
+                        "Token is sent to via email to the user : " + tenantAwareUsername);
+            }
+        } catch (NoSuchAlgorithmException e) {
+            throw new TOTPException(
+                    "TOTPTokenGenerator can't find the configured hashing algorithm", e);
+        } catch (InvalidKeyException e) {
+            throw new TOTPException("Secret key is not valid", e);
+        } catch (AuthenticationFailedException e) {
+            throw new TOTPException(
+                    "TOTPTokenVerifier cannot find the property value for encodingMethod");
+        }
+        return token;
+    }
+
+    /**
+     * Extract the email value from federated user attributes.
+     *
+     * @param userAttributes Map with federated user attributes.
+     * @return Email attribute.
+     */
+    private static String getEmailForFederatedUser(Map<ClaimMapping, String> userAttributes) {
+
+        String email = null;
+        for (Map.Entry<ClaimMapping, String> entry : userAttributes.entrySet()) {
```

Comment:
can userAttributes be null? if so handle null check

Reply:
Fixed